pkg/capacity/controller.go

```diff
@@ -0,0 +1,506 @@
+	} else {
+		c.queue.Forget(obj)
+	}
+
+	return true
+}
+
+// syncCapacity gets the capacity and then updates or creates the object.
+func (c *Controller) syncCapacity(ctx context.Context, item workItem) error {
+	// We lock only while accessing c.capacities, but not during
+	// the potentially long-running operations. That is okay
+	// because there is only a single worker per item. In the
+	// unlikely case that the desired state of the item changes
+	// while we work on it, we will add or update an obsolete
+	// object which we then don't store and instead queue for
+	// removal.
+	c.capacitiesLock.Lock()
+	capacity, found := c.capacities[item]
+	c.capacitiesLock.Unlock()
+
+	klog.V(5).Infof("Capacity Controller: refreshing %+v", item)
+	if !found {
+		// The item was removed in the meantime. This can happen when the storage class
+		// or the topology segment are gone.
+		klog.V(5).Infof("Capacity Controller: %v became obsolete", item)
+		return nil
+	}
+
+	sc, err := c.scInformer.Lister().Get(item.storageClassName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			// Another indication that the value is no
+			// longer needed.
+			return nil
+		}
+		return fmt.Errorf("retrieve storage class for %+v: %v", item, err)
+	}
+
+	req := &csi.GetCapacityRequest{
+		Parameters: sc.Parameters,
+		// The assumption is that the capacity is independent of the
+		// capabilities. The standard makes it mandatory to pass something,
+		// therefore we pick something rather arbitrarily.
+		VolumeCapabilities: []*csi.VolumeCapability{
+			{
+				AccessType: &csi.VolumeCapability_Mount{},
+				AccessMode: &csi.VolumeCapability_AccessMode{
+					Mode: csi.VolumeCapability_AccessMode_UNKNOWN,
+				},
+			},
+		},
+	}
+	if item.segment != nil {
+		req.AccessibleTopology = &csi.Topology{
+			Segments: item.segment.GetLabelMap(),
+		}
+	}
+	resp, err := c.csiController.GetCapacity(ctx, req)
+	if err != nil {
+		return fmt.Errorf("CSI GetCapacity for %+v: %v", item, err)
+	}
+
+	quantity := resource.NewQuantity(resp.AvailableCapacity, resource.BinarySI)
+	if capacity != nil {
+		// Update existing object.
+		capacity.Capacity = quantity
+		var err error
+		klog.V(5).Infof("Capacity Controller: updating %s for %+v, new capacity %v", capacity.Name, item, quantity)
+		capacity, err = c.client.StorageV1alpha1().CSIStorageCapacities().Update(ctx, capacity, metav1.UpdateOptions{})
+		if err != nil && apierrs.IsConflict(err) {
+			// Handle the case where we had a stale copy of the object. Can only happen
+			// when someone else was making changes to it, which should be rare.
+			capacity, err = c.client.StorageV1alpha1().CSIStorageCapacities().Get(ctx, capacity.Name, metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("getting fresh copy of CSIStorageCapacity for %+v: %v", item, err)
+			}
+			capacity.Capacity = quantity
+			capacity, err = c.client.StorageV1alpha1().CSIStorageCapacities().Update(ctx, capacity, metav1.UpdateOptions{})
```

Comment:
We usually propagate `Conflict` errors back to the workqueue, it gets retried after exp. backoff or when event that CSIStorageCapacity was updated arrives (if you had informer on CSIStorageCapacity...)

Or there is `RetryOnConflict()` if you really want to handle it here. https://github.com/kubernetes/client-go/blob/6f3bc4ba9215358ea3ceff566cff7480a5d23a7b/util/retry/util.go#L68

Reply:
At the moment, the CSIStorageCapacity object never gets updated in memory because there is no informer, so I have to deal with it here. `RetryOnConflict` looks like a simpler way to do it. Let me think about it this.

Reply:
I added the informer, so this special case could be removed.